compiler/src/iree/compiler/Dialect/HAL/IR/HALOpFolders.cpp

```diff
@@ -298,6 +300,256 @@ void CommandBufferPushDescriptorSetOp::getCanonicalizationPatterns(
+
+      // Branch from the header block into the first block of the region. Note
+      // that it may have a %device argument.
+      Block *firstBlock = headerBlock->getNextNode();
+      SmallVector<Value> firstBranchOperands;
+      if (firstBlock->getNumArguments() > 0) {
+        firstBranchOperands.push_back(newBlockOp.getArgument(0));
+      }
+      OpBuilder::atBlockEnd(headerBlock)
+          .create<cf::BranchOp>(newBlockOp.getLoc(), firstBlock,
+                                firstBranchOperands);
+
+      // Find the single expected return, capture its operands, and rewrite it
+      // to branch to the next block.
+      for (auto returnOp : llvm::make_early_inc_range(
+               targetRegion.getOps<IREE::HAL::ReturnOp>())) {
+        llvm::append_range(resultValues, returnOp.getOperands());
+        OpBuilder(returnOp).create<cf::BranchOp>(returnOp.getLoc(), nextBlock);
+        returnOp.erase();
+      }
+    }
+
+    // Return from the constant block with all operands.
+    OpBuilder::atBlockBegin(postBlock).create<IREE::HAL::ReturnOp>(
+        fusedLoc, resultValues);
+
+    rewriter.finalizeRootUpdate(variantOp);
+
+    // Erase all the old blocks.
+    for (auto blockOp : blockOps) {
+      rewriter.eraseOp(blockOp);
+    }
+
+    return success();
+  }
+};
+
+}  // namespace
+
+void ExecutableVariantOp::getCanonicalizationPatterns(
+    RewritePatternSet &results, MLIRContext *context) {
+  results.insert<MergeExecutableConstantBlocks>(context);
+}
+
+namespace {
+
+static void filterReturnOperands(ExecutableConstantBlockOp blockOp,
+                                 const BitVector &preservedIndices) {
+  for (auto returnOp :
+       llvm::make_early_inc_range(blockOp.getOps<IREE::HAL::ReturnOp>())) {
+    SmallVector<Value> operands;
+    for (auto [i, operand] : llvm::enumerate(returnOp.getOperands())) {
+      if (preservedIndices.test(i)) operands.push_back(operand);
+    }
+    returnOp.operandsMutable().assign(operands);
+  }
+}
+
+/// Drops the %device argument of a constant block region if unused.
+struct DropUnusedExecutableConstantBlockDeviceArg
+    : public OpRewritePattern<ExecutableConstantBlockOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(ExecutableConstantBlockOp blockOp,
+                                PatternRewriter &rewriter) const override {
+    if (blockOp.getNumArguments() == 0) return failure();
+    auto deviceArg = blockOp.getArgument(0);
+    if (!deviceArg.use_empty()) return failure();
+    rewriter.updateRootInPlace(blockOp, [&]() {
+      blockOp.eraseArgument(0);
+      blockOp.setFunctionTypeAttr(TypeAttr::get(
+          rewriter.getFunctionType(/*inputs=*/{}, blockOp.getResultTypes())));
+    });
+    return success();
+  }
+};
+
+/// Deduplicates constant values that have matching keys, choosing the first
+/// one found. There's no verification that the values produced are the same
+/// as users are expected to uniquely name their keys.
```

Comment:
Ah, I see here (from my other comment). I wonder if the test there could elaborate on its comment, or if some ops other than `arith.constant` would make the deduplication intent more clear.

Reply:
I can add a comment in the test and try to tweak the numbers to be more clear - I'm not sure additional IR would help

Reply:
Yeah, I had something like `test.sink` ops that we use in a few other tests in mind (`test.query_property()`?)